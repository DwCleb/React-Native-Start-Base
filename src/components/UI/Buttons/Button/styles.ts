import styled from 'styled-components/native';

import { Button } from 'react-native-paper';

import { Metrics, Colors } from '~/theme';
import { ThemeType } from '~/theme/theme';

export enum ButtonMode {
  TEXT = 'text',
  OUTLINED = 'outlined',
  CONTAINED = 'contained',
}

export interface ButtonProps {
  title: string;
  onPress(): void;
  accessibilityLabel?: string;
  icon?: string;
  color?: string;
  mode?: ButtonMode;
  loading?: boolean;
  disabled?: boolean;
  uppercase?: boolean;
  small?: boolean;
  distance?: number[];
  ratio?: number;
}

export type Props = ButtonProps & { theme: ThemeType };

function getMargin({ distance }: Props): string {
  if (!distance) {
    return `${10}px ${5}px ${10}px ${5}px`;
  } else {
    switch (distance.length || 0) {
      case 1:
        return `${distance[0]}px`;
      case 2:
        return `${distance[0]}px ${distance[1]}px`;
      case 3:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px`;
      case 4:
        return `${distance[0]}px ${distance[1]}px ${distance[2]}px ${distance[3]}px`;
      default:
        return '0px';
    }
  }
}

export const MyButton = styled(Button).attrs(
  ({
    icon,
    color = Colors.palette.primary.blue,
    mode = ButtonMode.CONTAINED,
    uppercase = false,
    loading,
    ratio = 0.85,
  }: ButtonProps) => ({
    icon,
    color,
    mode,
    uppercase,
    contentStyle: { height: 40, width: Metrics.screen.width * ratio },
    loading,
  })
)`
  margin: ${(props: Props): string => getMargin(props)};
  justify-content: center;
  align-items: center;
`;
