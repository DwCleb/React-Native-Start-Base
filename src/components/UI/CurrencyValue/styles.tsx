import Typography from '../Typography';
import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';

interface WrapperProps {
  color?: string;
  opacity: string;
  valueWeight: string;
  isCurrency?: boolean;
}

type ThemedProps = ThemeContextType & WrapperProps;

function getTitleColor({ theme, color }: ThemedProps): string {
  return color ? theme.currency.title[color] : theme.monochromatic.white;
}

export const LineTroughContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
  position: relative;
  margin-right: ${scale(3)}px;
`;

export const PercentageContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export const Value = styled(Typography)`
  color: ${(props: ThemedProps): string => `${getTitleColor(props)}${props.opacity === 'half' ? '7F' : ''}`};
  ${({ valueWeight }: ThemedProps): string => valueWeight && `font-weight: ${valueWeight}`};
  letter-spacing: ${(props: ThemedProps): string => (props.isCurrency ? `${scale(1.5)}` : `${scale(1)}`)}px;
`;

export const LineThrough = styled.View`
  width: 100%;
  height: ${scale(2)}px;
  background-color: yellow;
  position: absolute;
  bottom: auto;
`;

function padding(prop: number[] = []): string {
  switch (prop.length || 0) {
    case 1:
      return `${scale(prop[0])}px`;
    case 2:
      return `${scale(prop[0])}px ${scale(prop[1])}px`;
    case 3:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px`;
    case 4:
      return `${scale(prop[0])}px ${scale(prop[1])}px ${scale(prop[2])}px ${scale(prop[3])}px`;
    default:
      return '0px';
  }
}

export const Container = styled.View`
  flex-direction: ${({ row }: { row: string }): string => (row ? 'row' : 'column')};
  align-items: center;
  justify-content: flex-start;
  padding: ${({ distance }: { distance: [] }): string => padding(distance)};
`;
