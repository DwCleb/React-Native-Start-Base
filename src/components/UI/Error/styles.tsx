import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';
import Typography, { TypographyType } from '~/components/UI/Typography';
import Modal from 'react-native-modal';

//Modal
export const ErrorModal = styled(Modal)`
  margin: 0px;
`;

//Parent Component
export const MainContainer = styled.SafeAreaView`
  flex: 1;
  width: 100%;
  background-color: ${({ theme }: ThemeContextType): string => theme.background};
`;

//Content's ScrollView
export const Scroll = styled.ScrollView.attrs(() => ({
  keyboardShouldPersistTaps: 'handled',
  showsVerticalScrollIndicator: false,
  contentContainerStyle: { flex: 1, paddingTop: scale(35) },
}))`
  height: auto;
`;

//Content Container
export const ContentContainer = styled.View`
  flex: 1;
  margin: ${scale(30)}px;
`;

//Page Title
export const Title = styled(Typography).attrs(() => ({
  type: TypographyType.H2,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  margin-bottom: ${scale(10)}px;
  letter-spacing: ${scale(1)}px;
`;

export const SubTitle = styled(Typography).attrs(() => ({
  type: TypographyType.H5,
  weight: 'regular',
  align: 'left',
}))`
  color: ${({ theme }: ThemeContextType): string => theme.text.black};
  letter-spacing: ${scale(1)}px;
`;

//Error message
export const ErrorMessageContainer = styled.View`
  border-radius: ${scale(8)}px;
  height: ${scale(100)}px;
  padding: ${scale(20)}px ${scale(40)}px;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }: ThemeContextType): string => theme.error.card.background};
`;

export const ErrorMessage = styled(Typography).attrs(() => ({
  type: TypographyType.P2,
  weight: 'regular',
  align: 'center',
  numberOfLines: 2,
}))`
  color: ${({ theme }: ThemeContextType): string => theme.monochromatic.black};
  letter-spacing: ${scale(1)}px;
`;

//Buttons container
export const ButtonsContainer = styled.View`
  margin-bottom: ${scale(30)}px;
`;
