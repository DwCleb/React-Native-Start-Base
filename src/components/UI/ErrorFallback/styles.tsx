import styled from 'styled-components/native';
import { scale } from '~/theme';

import { Logo } from '~/assets';

export const ErrorImage = styled.Image.attrs(() => ({
  source: Logo,
  resizeMode: 'contain',
  align: 'center',
  marginTop: scale(30),
  marginBottom: scale(30),
  alignSelf: 'center',
}))`
  width: ${scale(82)}px;
`;
