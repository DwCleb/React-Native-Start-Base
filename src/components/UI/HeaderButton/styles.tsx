import styled from 'styled-components/native';

/** Theme */
import { scale, ThemeContextType } from '~/theme';

/** Assets */
import { BackIcon as BackIconAsset, CrossIcon as CrossIconAsset } from '~/assets';

/* Dependecies */
import { I18nManager } from 'react-native';

enum ButtonEnum {
  Back = 'back',
  Close = 'close',
}

interface IconProps {
  type: ButtonEnum;
}

export const ButtonContainer = styled.TouchableOpacity.attrs(() => ({
  hitSlop: { top: 30, left: 30, bottom: 30, right: 30 },
}))`
  margin-top: 0px;
  margin-left: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(20)}px` : '0px')};
  padding-right: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(30)}px` : '0px')};
`;

export const BackIcon = styled.Image.attrs((props: IconProps) => ({
  source: props.type === ButtonEnum.Back ? BackIconAsset : CrossIconAsset,
  resizeMode: 'contain',
}))`
  transform: scaleX(${I18nManager.isRTL ? -1 : 1});
  tint-color: ${(props: ThemeContextType): string => `${props.theme.navigation.back}`};
  height: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(16)}px` : `${scale(26)}px`)}
  width: ${(props: IconProps): string => (props.type === ButtonEnum.Back ? `${scale(9)}px` : `${scale(26)}px`)}
`;
