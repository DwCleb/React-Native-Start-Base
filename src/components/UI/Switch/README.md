```jsx
// TODO: create a Wrapper to react-styleguidist to provide theme to all components shown
import ThemeManager from '../../../../theme/ThemeContext';
import Switch from '.';

<ThemeManager>
  <Switch acessibilityLabel={'label true'} onChange={() => {}} value={true} />
  <Switch acessibilityLabel={'label false'} onChange={() => {}} value={false} />
</ThemeManager>;
```
