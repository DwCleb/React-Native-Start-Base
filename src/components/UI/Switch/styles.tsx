import styled from 'styled-components/native';
import { scale, ThemeContextType } from '~/theme';
import Typography, { TypographyType } from '~/components/UI/Typography';
import { View } from 'react-native';

interface Props {
  isEnabled: boolean;
  variant: 'primary' | 'secondary';
}

type ThemedProps = Props & ThemeContextType;

export const SwitchContainer = styled(View)`
  flex-direction: row;
  align-items: center;
`;

export const SwitchTrack = styled(View)`
  background-color: ${({ theme, variant, isEnabled }: ThemedProps): string =>
    isEnabled ? theme.switch.track[variant].active : theme.switch.track[variant].inactive};
  height: ${scale(18)}px;
  border-radius: ${scale(16)}px;
  align-items: ${({ isEnabled }: ThemedProps): string => (isEnabled ? 'flex-end' : 'flex-start')};
  justify-content: center;
  width: ${scale(26)}px;
`;

export const SwitchThumb = styled.View`
  height: ${scale(12)}px;
  width: ${scale(12)}px;
  border-radius: ${scale(12 / 2)}px;
  background-color: ${({ isEnabled, theme, variant }: ThemedProps): string =>
    isEnabled ? theme.switch.thumb[variant].active : theme.switch.thumb[variant].inactive};
  margin-left: ${scale(3)}px;
  margin-right: ${scale(3)}px;
`;

export const SwitchLabel = styled(Typography).attrs(() => ({
  type: TypographyType.L2,
  bold: true,
  uppercase: true,
}))`
  margin-left: ${scale(15)}px;
`;
