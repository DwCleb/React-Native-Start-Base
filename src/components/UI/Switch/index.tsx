import React from 'react';
import { TouchableHighlight } from 'react-native';

import { SwitchTrack, SwitchThumb, SwitchLabel, SwitchContainer } from './styles';

export enum SwitchVariantsEnum {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
}

export interface SwitchProps {
  /**  value that controls the switch */
  value: boolean;
  /**  function to change the value of the switch */
  onChange: (isEnabled: boolean) => void;
  /** label that can be displayed alongside the switch */
  acessibilityLabel?: string;
  /** color variants for the switch  */
  variant?: SwitchVariantsEnum;
  /** switch on press can be disabled */
  disabled?: boolean;
}

const Switch = ({
  value,
  onChange,
  acessibilityLabel,
  variant = SwitchVariantsEnum.PRIMARY,
  disabled,
}: SwitchProps): JSX.Element => {
  const handleChange = (newValue: boolean): void => {
    onChange(newValue);
  };

  return (
    <SwitchContainer>
      <TouchableHighlight onPress={(): void => handleChange(!value)} disabled={disabled}>
        <SwitchTrack isEnabled={value} variant={variant}>
          <SwitchThumb isEnabled={value} variant={variant} />
        </SwitchTrack>
      </TouchableHighlight>
      {acessibilityLabel && <SwitchLabel>{acessibilityLabel}</SwitchLabel>}
    </SwitchContainer>
  );
};

export default Switch;
