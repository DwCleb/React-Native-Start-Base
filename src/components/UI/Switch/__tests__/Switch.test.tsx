import React from 'react';

import { create, ReactTestRenderer, act } from 'react-test-renderer';

import Switch, { SwitchProps, SwitchVariantsEnum } from '~/components/UI/Controls/Switch';

import { ThemeProvider } from 'styled-components/native';

import lightTheme from '~/theme/ThemeContext/light';

import * as themeManager from '~/theme/ThemeContext';
import { TouchableHighlight } from 'react-native';
import { SwitchTrack, SwitchThumb, SwitchLabel } from '../styles';

const themeContext = jest.spyOn(themeManager, 'useTheme');

const theme = lightTheme.theme;

// FIXME: make it transversal to all tests
themeContext.mockImplementation(
  (): themeManager.ThemeContextType => {
    return {
      mode: 'light',
      setMode: jest.fn(),
      theme,
    };
  }
);

const onChangeEvent = jest.fn();

let wrapper: ReactTestRenderer;
let root: ReactTestRenderer['root'];

const props: SwitchProps = {
  value: false,
  onChange: onChangeEvent,
};

const componentToTest = (componentProps: SwitchProps): JSX.Element => (
  <ThemeProvider theme={theme}>
    <Switch {...componentProps} />
  </ThemeProvider>
);

describe('Render of Switch with mandatory props', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  beforeEach(() => {
    wrapper = create(componentToTest(props));
    root = wrapper.root;
  });

  it('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render the correct structure with default props', () => {
    /*renders pressable component */
    expect(root.findAllByType(TouchableHighlight)).toHaveLength(1);
    /*renders switch track */
    expect(root.findAllByType(SwitchTrack)).toHaveLength(1);
    expect(root.findAllByType(SwitchTrack)[0].props.isEnabled).toBe(false);
    /* renders switch thumb */
    expect(root.findAllByType(SwitchThumb)).toHaveLength(1);
    expect(root.findAllByType(SwitchThumb)[0].props.isEnabled).toBe(false);
  });

  it('Should change isEnabled prop after press on TouchableHighlight', () => {
    const ButtonElement = root.findAllByType(TouchableHighlight)[0];

    act(() => {
      ButtonElement.props.onPress();
    });

    expect(onChangeEvent).toHaveBeenCalled();
  });

  it('Should render components with isEnabled prop as true, if value prop is true', () => {
    const valueProps: SwitchProps = {
      ...props,
      value: true,
    };

    wrapper = create(componentToTest(valueProps));
    root = wrapper.root;

    const SwitchThumbElement = root.findAllByType(SwitchThumb)[0];
    const SwitchTrackElement = root.findAllByType(SwitchTrack)[0];
    //isEnabled prop is now true
    expect(SwitchThumbElement.props.isEnabled).toBe(true);
    expect(SwitchTrackElement.props.isEnabled).toBe(true);
  });

  it('Should render a SwitchLabel with acessibility label prop', () => {
    const labelProps: SwitchProps = {
      ...props,
      acessibilityLabel: 'label',
    };

    wrapper = create(componentToTest(labelProps));
    root = wrapper.root;

    /* renders label */
    const SwitchLabelElements = root.findAllByType(SwitchLabel);
    expect(SwitchLabelElements).toHaveLength(1);
    expect(SwitchLabelElements[0].props.children).toEqual(labelProps.acessibilityLabel);
  });

  it('Should render a Switch with secondary variant prop', () => {
    const variantProps: SwitchProps = {
      ...props,
      variant: SwitchVariantsEnum.SECONDARY,
    };

    wrapper = create(componentToTest(variantProps));
    root = wrapper.root;

    /* renders with the right variant */
    const SwitchTrackElement = root.findAllByType(SwitchTrack)[0];
    expect(SwitchTrackElement.props.variant).toEqual(variantProps.variant);
  });
});
