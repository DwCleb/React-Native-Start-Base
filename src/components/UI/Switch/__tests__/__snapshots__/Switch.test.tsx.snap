// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render of Switch with mandatory props Should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      isEnabled={false}
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#EEEEEE",
            "borderRadius": 32,
            "height": 36,
            "justifyContent": "center",
            "width": 52,
          },
        ]
      }
      variant="primary"
    >
      <View
        isEnabled={false}
        style={
          Array [
            Object {
              "backgroundColor": "#878A8F",
              "borderRadius": 12,
              "height": 24,
              "marginLeft": 6,
              "marginRight": 6,
              "width": 24,
            },
          ]
        }
        variant="primary"
      />
    </View>
  </View>
</View>
`;
